components/authz-service/engine/opa/opa_v2_internal_test.go

```diff
@@ -0,0 +1,988 @@
+			[2]*ast.Term{ast.NewTerm(ast.String("action")), ast.NewTerm(ast.String(action))},
+			[2]*ast.Term{ast.NewTerm(ast.String("projects")), ast.NewTerm(projs)},
+		)
+	}
+	result = r
+}
+
+// A: Specific input is faster!
+// BenchmarkV2GenericInput-8   	    377655	      4978 ns/op	    1664 B/op	      50 allocs/op
+// BenchmarkV2SpecificInput-8   	  540812	      2392 ns/op	     976 B/op	      30 allocs/op
+
+func BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies(b *testing.B) {
+	var r error
+	ctx := context.Background()
+
+	l, err := logger.NewLogger("text", "debug")
+	require.NoError(b, err, "init logger")
+	s, err := New(ctx, l)
+	require.NoError(b, err, "init state")
+
+	policyCounts := []int{0, 5, 10, 20, 50, 100, 200, 1000}
+	roleCount := 10 // keep this constant while increasing policyCount
+
+	chefPolicies, _ := v2BaselinePoliciesAndRoles()
+
+	for _, policyCount := range policyCounts {
+		policies, roles := v2BaselineAndRandomPoliciesAndRoles(policyCount, roleCount)
+		s.v2p1Store = inmem.NewFromObject(map[string]interface{}{
+			"policies": policies,
+			"roles":    roles,
+		})
+
+		b.Run(fmt.Sprintf("store with %d chef-managed policies and %d custom policies", len(chefPolicies), policyCount),
+			func(b *testing.B) {
+				for n := 0; n < b.N; n++ {
+					r = s.makeAuthorizedProjectPreparedQuery(ctx)
+					if r != nil {
+						b.Error(r)
+					}
+				}
+				errResult = r
+			})
+	}
+}
+
+// BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies/store_with_18_chef-managed_policies_and_0_custom_policies-8       12	  84996295 ns/op	 14671988 B/op	  373316 allocs/op
+// BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies/store_with_18_chef-managed_policies_and_5_custom_policies-8       12	 101366184 ns/op	 15695630 B/op	  394906 allocs/op
+// BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies/store_with_18_chef-managed_policies_and_10_custom_policies-8      13	 119584001 ns/op	 17528115 B/op	  433279 allocs/op
+// BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies/store_with_18_chef-managed_policies_and_20_custom_policies-8      10	 134156734 ns/op	 20404048 B/op	  493707 allocs/op
+// BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies/store_with_18_chef-managed_policies_and_50_custom_policies-8       8	 169178163 ns/op	 30520222 B/op	  704484 allocs/op
+// BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies/store_with_18_chef-managed_policies_and_100_custom_policies-8      4	 372269868 ns/op	 43306724 B/op	  975824 allocs/op
+// BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies/store_with_18_chef-managed_policies_and_200_custom_policies-8      2	 511771048 ns/op	 69466028 B/op	 1524912 allocs/op
+// BenchmarkAuthorizedProjectPreparedQueryWithIncreasingPolicies/store_with_18_chef-managed_policies_and_1000_custom_policies-8     1	1359604423 ns/op	257229008 B/op	 5492555 allocs/op
+// 12/16/19 summary: up to 1-2 seconds with over 1000 policies
```

Comment:
On L151, you just said "1000 teams", but here you start using "_over_ 1000 policies". Any reason to have the "over" ? (Applies to this benchmark and the next half dozen or so.)

Reply:
There's a 1000 custom policies + a few Chef-managed/system/migrated policies